regression - temporary value dropped while borrowed with static slice (as used by rust-phf)

The below code compiles in 1.40.0 but not in 1.42.0 (stable). (minimized example from rust-phf)

```rust
pub enum Slice<T: 'static> {
    Static(&'static [T]),
}

pub struct Map<K: 'static, V: 'static> {
    pub entries: Slice<(K, V)>,
}

static CONTENT : & 'static [ u8 ] = b"a";

pub static CONTENT_MAP: Map<&'static str, &'static [u8]> = {
    Map {
        entries: Slice::Static(&[
            ("content", CONTENT),
        ]),
    }
};
```

When built in 1.42.0, it produces the following error:

```shell
error[E0716]: temporary value dropped while borrowed
  --> src/lib.rs:13:33
   |
13 |            entries: Slice::Static(&[
   |   ________________________________-^
   |  |________________________________|
   | ||
14 | ||             ("content", CONTENT),
15 | ||         ]),
   | ||         ^
   | ||_________|
   | |__________creates a temporary which is freed while still in use
   |            cast requires that borrow lasts for `'static`
16 |        }
17 |    };
   |    - temporary value is freed at the end of this statement
```


The same error occurs using nightly.


There's an issue from January mentioning this build error on the rust-phf repo (seems to occur in 1.41.0+), but no followup discussion/work that I could find. https://github.com/sfackler/rust-phf/issues/187

--- a/src/librustc_mir/transform/promote_consts.rs
+++ b/src/librustc_mir/transform/promote_consts.rs
@@ -529,7 +529,9 @@ impl<'tcx> Validator<'_, 'tcx> {
                                         // https://github.com/rust-lang/rust/pull/74945#discussion_r463063247
                                         // There may be opportunity for generalization, but this needs to be
                                         // accounted for.
-                                        if proj_base.is_empty() && !self.tcx.is_thread_local_static(did) {
+                                        if proj_base.is_empty()
+                                            && !self.tcx.is_thread_local_static(did)
+                                        {
                                             not_promotable = false;
                                         }
                                     }

--- a/src/librustc_mir/transform/promote_consts.rs
+++ b/src/librustc_mir/transform/promote_consts.rs
@@ -506,6 +506,10 @@ impl<'tcx> Validator<'_, 'tcx> {
                 match *elem {
                     ProjectionElem::Deref => {
                         let mut not_promotable = true;
+                        // This is a special treatment for cases like *&STATIC where STATIC is a
+                        // global static variable.
+                        // This pattern is generated only when global static variables are directly
+                        // accessed and is qualified for promotion safely.
                         if let TempState::Defined { location, .. } = self.temps[local] {
                             let def_stmt =
                                 self.body[location.block].statements.get(location.statement_index);
@@ -517,7 +521,15 @@ impl<'tcx> Validator<'_, 'tcx> {
                             {
                                 if let Some(did) = c.check_static_ptr(self.tcx) {
                                     if let Some(hir::ConstContext::Static(..)) = self.const_kind {
-                                        if !self.tcx.is_thread_local_static(did) {
+                                        // The `is_empty` predicate is introduced to exclude the case
+                                        // where the projection operations are [ .field, * ].
+                                        // The reason is because promotion will be illegal if field
+                                        // accesses preceed the dereferencing.
+                                        // Discussion can be found at
+                                        // https://github.com/rust-lang/rust/pull/74945#discussion_r463063247 
+                                        // There may be opportunity for generalization, but this needs to be
+                                        // accounted for.
+                                        if proj_base.is_empty() && !self.tcx.is_thread_local_static(did) {
                                             not_promotable = false;
                                         }
                                     }

--- a/src/librustc_mir/transform/promote_consts.rs
+++ b/src/librustc_mir/transform/promote_consts.rs
@@ -502,9 +502,33 @@ impl<'tcx> Validator<'_, 'tcx> {
     fn validate_place(&self, place: PlaceRef<'tcx>) -> Result<(), Unpromotable> {
         match place {
             PlaceRef { local, projection: [] } => self.validate_local(local),
-            PlaceRef { local: _, projection: [proj_base @ .., elem] } => {
+            PlaceRef { local, projection: [proj_base @ .., elem] } => {
                 match *elem {
-                    ProjectionElem::Deref | ProjectionElem::Downcast(..) => {
+                    ProjectionElem::Deref => {
+                        let mut not_promotable = true;
+                        if let TempState::Defined { location, .. } = self.temps[local] {
+                            let def_stmt =
+                                self.body[location.block].statements.get(location.statement_index);
+                            if let Some(Statement {
+                                kind:
+                                    StatementKind::Assign(box (_, Rvalue::Use(Operand::Constant(c)))),
+                                ..
+                            }) = def_stmt
+                            {
+                                if let Some(did) = c.check_static_ptr(self.tcx) {
+                                    if let Some(hir::ConstContext::Static(..)) = self.const_kind {
+                                        if !self.tcx.is_thread_local_static(did) {
+                                            not_promotable = false;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        if not_promotable {
+                            return Err(Unpromotable);
+                        }
+                    }
+                    ProjectionElem::Downcast(..) => {
                         return Err(Unpromotable);
                     }
 

--- a/src/librustc_mir/transform/promote_consts.rs
+++ b/src/librustc_mir/transform/promote_consts.rs
@@ -526,7 +526,7 @@ impl<'tcx> Validator<'_, 'tcx> {
                                         // The reason is because promotion will be illegal if field
                                         // accesses preceed the dereferencing.
                                         // Discussion can be found at
-                                        // https://github.com/rust-lang/rust/pull/74945#discussion_r463063247 
+                                        // https://github.com/rust-lang/rust/pull/74945#discussion_r463063247
                                         // There may be opportunity for generalization, but this needs to be
                                         // accounted for.
                                         if proj_base.is_empty() && !self.tcx.is_thread_local_static(did) {

--- a/src/test/ui/statics/static-promotion.rs
+++ b/src/test/ui/statics/static-promotion.rs
@@ -12,12 +12,23 @@ struct A<T: 'static>(&'static T);
 struct B<T: 'static + ?Sized> {
     x: &'static T,
 }
+static STR: &'static [u8] = b"hi";
 static C: A<B<B<[u8]>>> = {
     A(&B {
-        x: &B { x: b"hi" as &[u8] },
+        x: &B { x: STR },
     })
 };
 
+pub struct Slice(&'static [i32]);
+
+static CONTENT: i32 = 42;
+pub static CONTENT_MAP: Slice = Slice(&[CONTENT]);
+
+pub static FOO: (i32, i32) = (42, 43);
+pub static CONTENT_MAP2: Slice = Slice(&[FOO.0]);
+
 fn main() {
     assert_eq!(b"hi", C.0.x.x);
+    assert_eq!(&[42], CONTENT_MAP.0);
+    assert_eq!(&[42], CONTENT_MAP2.0);
 }

--- a/src/test/ui/statics/static-promotion.rs
+++ b/src/test/ui/statics/static-promotion.rs
@@ -1,4 +1,12 @@
-// run-pass
+// check-pass
+
+// Use of global static variables in literal values should be allowed for
+// promotion.
+// This test is to demonstrate the issue raised in
+// https://github.com/rust-lang/rust/issues/70584
+
+// Literal values were previously promoted into local static values when
+// other global static variables are used.
 
 struct A<T: 'static>(&'static T);
 struct B<T: 'static + ?Sized> {

--- a/src/test/ui/statics/static-promotion.rs
+++ b/src/test/ui/statics/static-promotion.rs
@@ -1,0 +1,15 @@
+// run-pass
+
+struct A<T: 'static>(&'static T);
+struct B<T: 'static + ?Sized> {
+    x: &'static T,
+}
+static C: A<B<B<[u8]>>> = {
+    A(&B {
+        x: &B { x: b"hi" as &[u8] },
+    })
+};
+
+fn main() {
+    assert_eq!(b"hi", C.0.x.x);
+}

